File: storm/src/main/java/utils/CoordinateHelper.java
package utils;

public class CoordinateHelper {
    private static final int EARTH_RADIUS = 6371000;

    /**
     * @return distance in meters
     */
    public static double calculateDistance(TaxiLog t1, TaxiLog t2) {
        double phi1 = (t1.getLatitude() / 360) * 2 * Math.PI;
        double phi2 = (t2.getLatitude() / 360) * 2 * Math.PI;
        double lambda1 = (t1.getLongitude() / 360) * 2 * Math.PI;
        double lambda2 = (t2.getLongitude() / 360) * 2 * Math.PI;

        return 2 * EARTH_RADIUS * Math.asin(
                Math.sqrt(Math.pow(Math.sin((Math.abs(phi2 - phi1))/2), 2) +
                        Math.cos(phi1) * Math.cos(phi2)*Math.pow(Math.sin((Math.abs(lambda2 - lambda1))/2),2)));
    }
}

File: dashboard-frontend/src/CoordinateHelper.ts
export const EARTH_RADIUS = 6371000;
export const LAT_BEIJING = 39.916668;
export const LONG_BEIJING = 116.383331;

export const calculateDistance = (lat1: number, long1: number, lat2: number, long2: number) => {
  let phi1 = (lat1 / 360) * 2 * Math.PI;
  let phi2 = (lat2 / 360) * 2 * Math.PI;
  let lambda1 = (long1 / 360) * 2 * Math.PI;
  let lambda2 = (long2 / 360) * 2 * Math.PI;

  return 2 * EARTH_RADIUS * Math.asin(
    Math.sqrt(Math.pow(Math.sin((Math.abs(phi2 - phi1))/2), 2) +
      Math.cos(phi1) * Math.cos(phi2)*Math.pow(Math.sin((Math.abs(lambda2 - lambda1))/2),2)));
}

export const calculateDistanceToBeijingCenter = (lat1: number, long1: number) => {
  return calculateDistance(lat1, long1, LAT_BEIJING, LONG_BEIJING);
}